render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java

```diff
@@ -30,37 +35,101 @@ public ResolvedTileSpecsWithMatchPairs(final ResolvedTileSpecCollection resolved
         this.resolvedTileSpecs = resolvedTileSpecs;
         this.matchPairs = matchPairs;
-        this.isNormalized = false;
     }
 
     public ResolvedTileSpecCollection getResolvedTileSpecs() {
-        if (! isNormalized) {
-            resolveTileSpecsAndNormalizeMatchPairs();
-        }
         return resolvedTileSpecs;
     }
 
+    @JsonIgnore
+    public TileSpec getTileSpec(final String tileId) {
+        return resolvedTileSpecs.getTileSpec(tileId);
+    }
+
     public List<CanvasMatches> getMatchPairs() {
-        if (! isNormalized) {
-            resolveTileSpecsAndNormalizeMatchPairs();
-        }
         return matchPairs;
     }
 
-    private synchronized void resolveTileSpecsAndNormalizeMatchPairs()
+    @JsonIgnore
+    public int getMatchPairCount() {
+        return matchPairs.size();
+    }
+
+    /**
+     * Normalizes this collection using the specified parameters by resolving all tile specs for client-side usage,
+     * filtering out unwanted tiles, removing match pairs that are too far from each other in z, and sorting pairs.
+     *
+     * @param  tileIdsToKeep  set of tileIds to retain in the collection (or null to retain all tiles).
+     *                        Match pairs for any removed tiles will also be removed.
+     * @param  maxZDistance   maximum non-negative integral z distance for all retained pairs
+     *                        (or null to accept all pairs).
+     *
+     * @throws IllegalArgumentException
+     *   if maxZDistance < 0
+     */
+    public void normalize(final Set<String> tileIdsToKeep,
+                          final Integer maxZDistance)
             throws IllegalArgumentException {
-        if (! isNormalized) {
-            resolvedTileSpecs.resolveTileSpecs();
-            final List<CanvasMatches> normalizedMatchPairs = new ArrayList<>(matchPairs.size());
-            for (final CanvasMatches pair : matchPairs) {
-                if (resolvedTileSpecs.hasTileSpec(pair.getpId()) && (resolvedTileSpecs.hasTileSpec(pair.getqId()))) {
+
+        final Integer numberOfTilesToKeep = tileIdsToKeep == null ? null : tileIdsToKeep.size();
+
+        LOG.info("normalize: entry, process {} tiles and {} pairs with {} tileIdsToKeep and maxZDistance {}",
+                 resolvedTileSpecs.getTileCount(), matchPairs.size(), numberOfTilesToKeep, maxZDistance);
+
+        if ((maxZDistance != null) && (maxZDistance <= 0)) {
+            throw new IllegalArgumentException("maxZDistance must be >= 0 or null");
+        }
+
+        // remove unwanted tiles (which will later cause associated math pairs to be removed)
+        if (tileIdsToKeep != null) {
+
+            // track and log removal info if debugging
+            final boolean isDebugEnabled = LOG.isDebugEnabled();
+            final Set<String> beforeTileIds = isDebugEnabled ? new HashSet<>(resolvedTileSpecs.getTileIds()) : null;
+
+            resolvedTileSpecs.removeDifferentTileSpecs(tileIdsToKeep);
```

Comment:
The 'standard name' for this operation in the `Collection` interface would be `retainAll()`.

Reply:
Yeah - I did not realize that years ago when I wrote the original method - doh.  I've updated the name.